In [1]:
# https://www.kaggle.com/c/mercedes-benz-greener-manufacturing/data?select=sample_submission.csv.zip

In [1]:
import os
import pandas as pd

os.chdir("E:/code/6.MachineLearning/5_Preprocessing_of_data")
df = pd.read_csv("https://raw.githubusercontent.com/pavanpuppala0346/6.MachineLearning/main/5_Preprocessing_of_data/mercedes_benz.csv",usecols=["X1","X2","X3","X4","X5","X6"])
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [2]:
df.shape

(4209, 6)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   X1      4209 non-null   object
 1   X2      4209 non-null   object
 2   X3      4209 non-null   object
 3   X4      4209 non-null   object
 4   X5      4209 non-null   object
 5   X6      4209 non-null   object
dtypes: object(6)
memory usage: 197.4+ KB


In [4]:
# how many unique values in each col
for col in df.columns:
    print(col,":",len(df[col].unique()),"labels")

X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels


In [5]:
# if we perform one hot encoding using get_dummies
import pandas as pd

pd.get_dummies(df, drop_first=True).shape

(4209, 117)

- from 6 categorical columns we end up with 117 columns
- if the columns increases it will impacts on accuracy so now we have to fix this
- To overcome this one hot encoding to the 10 most frequently labels of the variable. this means that they would make one binary variable for each of 10 most frequently labels only. this is equivalent to grouping all other variables under a new category, that in this case we will dropped. thus the 10 new dummy variables inidicate if one of the 10 most frequent labels is present (1) or (0) for a particular observation

In [6]:
# find the top 10 most frquent categories for the variable X2

df.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: X2, dtype: int64

In [7]:
# top10 = [x for x in df.X2.value_counts().sort_values(ascending=False).head(10).index]
top10 = df.X2.value_counts().sort_values(ascending=False).head(10).index.to_list()
top10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [8]:
# we make 10 binary variables
import numpy as np

for label in top10:
    df[label] = np.where(df["X2"]==label,1,0)

In [9]:
df[["X2"]+top10].head(20)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0

In [10]:
def one_hot_encoding(df, variable, top_x_labels):
    for label in top_x_labels:
        df[variable+"-"+label] = np.where(df[variable]==label, 1, 0)

In [11]:
df = pd.read_csv("mercedes_benz.csv", usecols=["X1","X2","X3","X4","X5","X6"])
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [12]:
one_hot_encoding(df, "X2", top10)

In [13]:
df

X1  X2 X3 X4  X5 X6  X2-as  X2-ae  X2-ai  X2-m  X2-ak  X2-r  X2-n  X2-s  \
0     v  at  a  d   u  j      0      0      0     0      0     0     0     0   
1     t  av  e  d   y  l      0      0      0     0      0     0     0     0   
2     w   n  c  d   x  j      0      0      0     0      0     0     1     0   
3     t   n  f  d   x  l      0      0      0     0      0     0     1     0   
4     v   n  f  d   h  d      0      0      0     0      0     0     1     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...    ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0      0     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0      0     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0      0     1     0     0   
4207  r   e  f  d  aa  l      0      0      0     0      0     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0      0     0     0     0   

      X2-f  X2-e  
0        0     0  
1        0     0  
2        0     0  
3        0     0  
4        0     0  
...    ...   ...  
4204     0     0  
4205     0     0  
4206     0     0  
4207     0     1  
4208     0     0  

[4209 rows x 16 columns]